File: Yandex/Practicum/Algorightms_and_Data_Structures/2_sprint/F.Стек-Max/solution.cpp
#include <iostream>
#include <string>
#include <vector>

class Stack {
private:
    std::vector<int> stack;
    std::vector<int> maxStack;
public:
    Stack() {};
    void getMax() {
        if (stack.size()) {
            std::cout << maxStack[maxStack.size() - 1] << "\n";
            return;
        }
        std::cout << "None\n";
    }
    void push(int x) {
        maxStack.push_back((stack.size()? std::max(x, maxStack[maxStack.size() - 1]) : x));
        stack.push_back(x);
    }
    void pop() {
        if (stack.size()) {
            maxStack.pop_back();
            stack.pop_back();
            return;
        }
        std::cout << "error\n";
    }
};
void solution(int t) {
    /*
    Time Complexity O(1)
    Memory Complexity O(N)
    */
    Stack stack = Stack();
    for (int i = 0; i < t; ++i) {
        std::string method;
        std::cin >> method;
        if (method == "get_max") {
            stack.getMax();
        } else if (method == "push") {
            int num;
            std::cin >> num;
            stack.push(num);
        } else {
            stack.pop();
        }
    }
}


int main() {
    int t;
    std::cin >> t;
    solution(t);
}

File: Yandex/Practicum/Algorightms_and_Data_Structures/2_sprint/F.Стек-Max/solution.py
import sys


class Stack:
    def __init__(self) -> None:
        self.stack: list[int] = []
        self.max_stack: list[int] = []

    def pop(self) -> int:
        if self.stack:
            self.max_stack.pop()
            return self.stack.pop()
        sys.stdout.write('error\n')

    def push(self, x: int) -> None:
        self.max_stack.append(x if not self.stack else max(x, self.max_stack[-1]))
        self.stack.append(x)

    def get_max(self) -> int | str:
        return self.max_stack[-1] if self.stack else 'None'


def solution(t: int) -> None:
    """
    Time Complexity O(1)
    Memory Complexity O(N)
    """
    stack: Stack = Stack()
    for _ in range(t):
        method: str = sys.stdin.readline().rstrip()
        if method.startswith('get_max'):
            sys.stdout.write(str(stack.get_max()) + '\n')
        elif method.startswith('push'):
            stack.push(int(method.split()[-1]))
        else:
            stack.pop()


if __name__ == '__main__':
    t: int = int(sys.stdin.readline().rstrip())
    solution(t)
